### PR TITLE
[full-ci][tests-only] use global cache for web and core repos usage

### PR DESCRIPTION
**Pull request by kiranparajuli589**
Signed-off-by: Kiran Parajuli <kiranparajuli589@gmail.com>

## Description
- `curl` to check if the cache already exists in the cache server
- if found, builds use the existing cache (only restore)
- if not found, a new cache for the commit id is pushed to the cache server (build + restore)

Cached:
- web
- yarn for web e2e tests
- yarn for acceptance tests

## Related Issue
- Part of https://github.com/owncloud/QA/issues/756

## How Has This Been Tested?
- :robot:

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [x] Tests only (no source changes)

### CODE REVIEW

**Comment by ownclouders** [comment]
:boom: Acceptance test [<strong>cs3ApiTests-ocis</strong>](https://drone.owncloud.com/owncloud/ocis/13999/40/1) failed. Further test are cancelled...

**Comment by kiranparajuli589** [comment]
Without changes in `.drone.env`: https://drone.owncloud.com/owncloud/ocis/13994
With changes in the `.drone.env`: https://drone.owncloud.com/owncloud/ocis/14001

File: .drone.star [hunk 1]
```diff
@@ -2801,3 +2764,82 @@ def setupForLitmus():
...
+                "MC_HOST": {
+                    "from_secret": "cache_s3_endpoint",
+                },
+                "AWS_ACCESS_KEY_ID": {
+                    "from_secret": "cache_s3_access_key",
+                },
+                "AWS_SECRET_ACCESS_KEY": {
+                    "from_secret": "cache_s3_secret_key",
+                },
+            },
```
Comment (grgprarup) [inline .drone.star:2784]:
```suggestion
            "environment": MINIO_MC_ENV,
```

File: .drone.star [hunk 2]
```diff
@@ -2801,3 +2773,137 @@ def setupForLitmus():
...
+                "mc alias set s3 $MC_HOST $AWS_ACCESS_KEY_ID $AWS_SECRET_ACCESS_KEY",
+                "mc mirror --overwrite --remove webTestRunner/.yarn s3/$CACHE_BUCKET/ocis/web-test-runner/yarn-e2e-$WEB_COMMITID",
+                "mc mirror --overwrite --remove webTestRunner/tests/acceptance/.yarn s3/$CACHE_BUCKET/ocis/web-test-runner/yarn-acceptance-$WEB_COMMITID",
+            ],
+        },
+    ]
+
+def restoreWebCache():
+    return [{
+        "name": "retore-web-cache",
```
Comment (grgprarup) [inline .drone.star:2857]:
```suggestion
        "name": "restore-web-cache",
```

File: .drone.star [hunk 3]
```diff
@@ -2801,3 +2773,137 @@ def setupForLitmus():
...
+            "mc alias set s3 $MC_HOST $AWS_ACCESS_KEY_ID $AWS_SECRET_ACCESS_KEY",
+            "mc mirror s3/$CACHE_BUCKET/ocis/web-test-runner/$WEB_COMMITID %s" % dirs["web"],
+            "ls -la %s" % dirs["web"],
+            "chmod +x %s/tests/acceptance/*.sh" % dirs["web"],
+        ],
+    }]
+
+def restoreWebE2EYarnCache():
+    return [{
+        "name": "retore-web-e2e-yarn-cache",
```
Comment (grgprarup) [inline .drone.star:2871]:
```suggestion
        "name": "restore-web-e2e-yarn-cache",
```

File: .drone.star [hunk 4]
```diff
@@ -2801,3 +2773,137 @@ def setupForLitmus():
...
+        "image": OC_CI_NODEJS % DEFAULT_NODEJS_VERSION,
+        "commands": [
+            "cd %s" % dirs["web"],
+            "retry -t 3 'yarn install --immutable'",
+        ],
+    }]
+
+def restoreWebAcceptanceYarnCache():
+    return [{
+        "name": "retore-web-acceptance-yarn-cache",
```
Comment (grgprarup) [inline .drone.star:2892]:
```suggestion
        "name": "restore-web-acceptance-yarn-cache",
```

**Review by grgprarup — COMMENTED** [review]
LGTM :+1:

**Review by SwikritiT — COMMENTED** [review]
LGTM

File: .drone.star [hunk 5]
```diff
@@ -2801,3 +2773,137 @@ def setupForLitmus():
...
+    path_to_drone_env = "%s/%s/.drone.env" % (ocis_git_base_url, ctx.build.commit)
+    path_to_check_script = "%s/%s/tests/config/drone/check_web_cache.sh" % (ocis_git_base_url, ctx.build.commit)
+
+    return [
+        {
+            "name": "get-drone-env-and-check-script",
+            "image": OC_UBUNTU,
+            "commands": [
+                "curl -s -o .drone.env %s" % path_to_drone_env,
+                "cat .drone.env",
```
Comment (individual-it) [inline .drone.star:2788]: do we need this debug line?

File: .drone.star [hunk 6]
```diff
@@ -2801,3 +2773,137 @@ def setupForLitmus():
...
+    ocis_git_base_url = "https://raw.githubusercontent.com/owncloud/ocis"
+    path_to_drone_env = "%s/%s/.drone.env" % (ocis_git_base_url, ctx.build.commit)
+    path_to_check_script = "%s/%s/tests/config/drone/check_web_cache.sh" % (ocis_git_base_url, ctx.build.commit)
+
+    return [
+        {
+            "name": "get-drone-env-and-check-script",
+            "image": OC_UBUNTU,
+            "commands": [
+                "curl -s -o .drone.env %s" % path_to_drone_env,
```
Comment (individual-it) [inline .drone.star:2787]: that file and also the script is in your clone, why downloading it again?

**Review by individual-it — CHANGES_REQUESTED** [review]

**Comment by individual-it** [comment]
please also run with full-ci

File: .drone.star (hunk 6 above)
Comment (kiranparajuli589) [reply .drone.star:2787]: for this pipeline we just need two files from ocis repo: 
1. `.drone.env` and 
2. `script to check for existing cache`

For only these two filea, the whole repo clone is not necessary. So repo clone is disabled in this pipeline. And these necessary files are downloaded using the `rawcontent` facility of github where we can grab file content from specific commit id

File: .drone.star (hunk 5 above)
Comment (kiranparajuli589) [reply .drone.star:2788]: nope

File: .drone.star [hunk 7]
```diff
@@ -2799,3 +2771,135 @@ def setupForLitmus():
...
+                "curl -s -o check_web_cache.sh %s" % path_to_check_script,
+                "chmod +x check_web_cache.sh",
```
Comment (individual-it) [inline .drone.star:2787]: I don't think you need this because you are using `bash check_web_cache.sh`, the executable is `bash` in that case

**Review by individual-it — CHANGES_REQUESTED** [review]

**Comment by phil-davis** [comment]
https://drone.owncloud.com/owncloud/ocis/14271/83/3
docker-amd64 "generate go" step fails.
`2022/08/09 08:57:27 Unexpected package creation during export data loading`
Is that a bug or a feature?

**Comment by kiranparajuli589** [comment]
More Logs 
```console
/go/bin/mockery-v2.10.4 --dir pkg/service/v0 --case underscore --name GatewayClient
09 Aug 22 08:57 UTC INF Starting mockery dry-run=false version=v2.10.4
09 Aug 22 08:57 UTC INF Walking dry-run=false version=v2.10.4
2022/08/09 08:57:27 Unexpected package creation during export data loading
make[1]: *** [Makefile:28: ci-go-generate] Error 1
make: *** [Makefile:152: ci-go-generate] Error 1
Retries exhausted
```

> restarting to see if it happens again

**Comment by kiranparajuli589** [comment]
> https://drone.owncloud.com/owncloud/ocis/14271/83/3 docker-amd64 "generate go" step fails. `2022/08/09 08:57:27 Unexpected package creation during export data loading` Is that a bug or a feature?

restarting to check if fails again.

File: .drone.star [hunk 8]
```diff
@@ -624,7 +589,7 @@ def localApiTests(ctx, storage, suite, accounts_hash_difficulty = 4):
                              "PATH_TO_CORE": "/drone/src/%s" % dirs["core"],
                              "TEST_SERVER_URL": "https://ocis-server:9200",
                              "OCIS_REVA_DATA_ROOT": "%s" % ("/srv/app/tmp/ocis/owncloud/data/" if storage == "owncloud" else ""),
-                             "SKELETON_DIR": "%s/data/apiSkeleton" % dirs["testing_app"],
+                             "SKELETON_DIR": "/drone/src/%s/data/apiSkeleton" % dirs["testing_app"],
```
Comment (phil-davis) [inline .drone.star:592]:
```suggestion
                             "SKELETON_DIR": "%s/%s/data/apiSkeleton" % (dirs["base"], dirs["testing_app"]),
```

File: .drone.star [hunk 9]
```diff
@@ -2799,3 +2772,134 @@ def setupForLitmus():
...
+                # cache using the minio/mc client to the 'owncloud' bucket (long term bucket)
+                "mc alias set s3 $MC_HOST $AWS_ACCESS_KEY_ID $AWS_SECRET_ACCESS_KEY",
+                "mc mirror --overwrite --remove webTestRunner s3/$CACHE_BUCKET/ocis/web-test-runner/$WEB_COMMITID",
+            ],
+        },
+        {
+            "name": "install-yarn",
+            "image": OC_CI_NODEJS % DEFAULT_NODEJS_VERSION,
+            "commands": [
+                "cd %s/webTestRunner" % dirs["base"],
```
Comment (phil-davis) [inline .drone.star:2831]:
```suggestion
                "cd %s" % dirs["web"],
```

File: .drone.star [hunk 10]
```diff
@@ -2799,3 +2772,134 @@ def setupForLitmus():
...
+                "retry -t 3 'yarn install --immutable --frozen-lockfile'",
+                "cd %s/webTestRunner/tests/acceptance" % dirs["base"],
```
Comment (phil-davis) [inline .drone.star:2833]:
```suggestion
                "cd %s/tests/acceptance" % dirs["web"],
```

File: .drone.star [hunk 11]
```diff
@@ -2799,3 +2772,134 @@ def setupForLitmus():
...
+            "commands": [
+                "bash check_web_cache.sh",
+            ],
+        },
+        {
+            "name": "clone-web",
+            "image": OC_CI_NODEJS % DEFAULT_NODEJS_VERSION,
+            "commands": [
+                ". ./.drone.env",
+                "rm -rf /drone/src/webTestRunner",
```
Comment (phil-davis) [inline .drone.star:2810]:
```suggestion
                "rm -rf %s" % dirs["web"],
```

File: .drone.star [hunk 12]
```diff
@@ -41,6 +41,8 @@ DEFAULT_PHP_VERSION = "7.4"
 DEFAULT_NODEJS_VERSION = "14"
 
 dirs = {
+    "base": "/drone/src",
+    "web": "/drone/src/webTestRunner",
```
Comment (phil-davis) [inline .drone.star:45]: There are quite a few hard-coded mentions of "webTestRunner".
Maybe we can have:
```suggestion
    "web_test_runner": "webTestRunner",
```
and use it like "core" and "testing_app" that are below - relative to "base". Then we can construct the base+web_test_runner path when needed, or just substitute dirs["web_test_runner"] when just that directory name is needed in a string.

File: .drone.star [hunk 13]
```diff
@@ -624,7 +589,7 @@ def localApiTests(ctx, storage, suite, accounts_hash_difficulty = 4):
                              "PATH_TO_CORE": "/drone/src/%s" % dirs["core"],
```
Comment (grgprarup) [inline .drone.star:589]:
```suggestion
                             "PATH_TO_CORE": "%s/%s" % (dirs["base"], dirs["core"]),
```

**Comment by kiranparajuli589** [comment]
More unrelated failures at
- https://drone.owncloud.com/owncloud/ocis/14283/7/3
    ```console
    level=error msg="Running error: 1 error occurred:\n\t* can't run linter goanalysis_metalinter: buildir: failed to load package : could not load export data: no export data for \"google.golang.org/genproto/googleapis/rpc/status\"\n\n"
    ```

- https://drone.owncloud.com/owncloud/ocis/14283/14/2
    ```console
    2022/08/09 10:16:51 Unexpected package creation during export data loading
    ```

File: .drone.star [hunk 14]
```diff
@@ -993,42 +965,43 @@ def settingsUITests(ctx, storage = "ocis", accounts_hash_difficulty = 4):
...
+                 [
+                     {
+                         "name": "WebUIAcceptanceTests",
+                         "image": OC_CI_NODEJS % DEFAULT_NODEJS_VERSION,
+                         "environment": {
+                             "SERVER_HOST": "https://ocis-server:9200",
+                             "BACKEND_HOST": "https://ocis-server:9200",
+                             "RUN_ON_OCIS": "true",
+                             "OCIS_REVA_DATA_ROOT": "/srv/app/tmp/ocis/owncloud/data",
+                             "WEB_UI_CONFIG": "/drone/src/tests/config/drone/ocis-config.json",
```
Comment (grgprarup) [inline .drone.star:984]:
```suggestion
                             "WEB_UI_CONFIG": "%s/tests/config/drone/ocis-config.json" % dirs["base"],
```

File: .drone.star [hunk 15]
```diff
@@ -993,42 +965,43 @@ def settingsUITests(ctx, storage = "ocis", accounts_hash_difficulty = 4):
...
+                             "TEST_TAGS": "not @skipOnOCIS and not @skip",
+                             "LOCAL_UPLOAD_DIR": "/uploads",
+                             "NODE_TLS_REJECT_UNAUTHORIZED": 0,
+                             "WEB_PATH": dirs["web"],
+                             "FEATURE_PATH": "/drone/src/services/settings/ui/tests/acceptance/features",
```
Comment (grgprarup) [inline .drone.star:989]:
```suggestion
                             "FEATURE_PATH": "%s/services/settings/ui/tests/acceptance/features" % dirs["base"],
```

File: .drone.star [hunk 16]
```diff
@@ -993,42 +965,43 @@ def settingsUITests(ctx, storage = "ocis", accounts_hash_difficulty = 4):
...
+                             "MIDDLEWARE_HOST": "http://middleware:3000",
+                         },
+                         "commands": [
+                             # TODO: settings/package.json has all the acceptance test dependencies
+                             # they shouldn't be needed since we could also use them from web:/tests/acceptance/package.json
+                             "cd /drone/src/services/settings",
```
Comment (grgprarup) [inline .drone.star:995]:
```suggestion
                             "cd %s/services/settings" % dirs["base"],
```

**Comment by kiranparajuli589** [comment]
@phil-davis I'd to use the relative path for core repos (core + testing). This is because of caching using the `PLUGINS_S3_CACHE` library. This does not support absolute urls and crashes in between if provided. So we've to use relative urls for this. Due to which we've to use `dirs[base], dir[core]` combinations for such usages

But this is not the same for web. Because we use `minio-mc`  library to cache web. This supports caching using absolute/relative any type of paths.

That's why I've two types of path in the dirs dictionary.

**Review by phil-davis — APPROVED** [review]

**Comment by phil-davis** [comment]
@individual-it please review again.

IMO we can refactor more old hard-coded mentions of `/drone/src` etc. in some other PR - that is not directly related to this change.

**Review by individual-it — APPROVED** [review]

**Comment by phil-davis** [comment]
> I've tried to avoid more hard-coded path usages in the last commit. But there are many such usages which might not be relevant to change in this same PR.

See PR #4372

**Pull request merged**